### PR TITLE
Disable eager activities for incompatible versioned activities

### PR DESCRIPTION
**Pull request by bergundy**
Only enable eager activity start if dynamic config enables it and workflow doesn't use versioning.
If a workflow _uses_ versioning, only allow eager activities which intend to use a compatible version since a
worker is obviously compatible with itself and we are okay dispatching an eager task knowning that there may be a
newer "default" compatible version.
Note that if `UseCompatibleVersion` is false, it implies that the activity should run on the "default" version
for the task queue.

### CODE REVIEW

File: service/history/workflow_task_handler.go [hunk 1]
```diff
@@ -384,7 +384,15 @@ func (handler *workflowTaskHandlerImpl) handleCommandScheduleActivity(
...
+	// Only enable eager activity start if dynamic config enables it and workflow doesn't use versioning.
+	// If a workflow _uses_ versioning, only allow eager activities which intend to use a compatible version since a
+	// worker is obviously compatible with itself and we are okay dispatching an eager task knowning that there may be a
+	// newer "default" compatible version.
+	// Note that if `UseCompatibleVersion` is false, it implies that the activity should run on the "default" version
+	// for the task queue.
+	if attr.RequestEagerExecution && handler.config.EnableActivityEagerExecution(namespace) &&
+		(!handler.mutableState.GetWorkerVersionStamp().GetUseVersioning() || attr.UseCompatibleVersion) {
 		eagerStartActivity = true
 	}
```
Comment (dnr) [inline service/history/workflow_task_handler.go:397]: how about set it once and remove line 385?
```suggestion
	eagerStartActivity := attr.RequestEagerExecution && handler.config.EnableActivityEagerExecution(namespace) &&
		(!handler.mutableState.GetWorkerVersionStamp().GetUseVersioning() || attr.UseCompatibleVersion)
```

File: tests/versioning_test.go [hunk 2]
```diff
@@ -798,6 +799,80 @@ func (s *versioningIntegSuite) dispatchActivityCompatible() {
...
+		},
+	})
+	s.Require().NoError(err)
+	startToCloseTimeout := time.Minute
+
+	completionResponse, err := s.sdkClient.WorkflowService().RespondWorkflowTaskCompleted(ctx, &workflowservice.RespondWorkflowTaskCompletedRequest{
+		Identity: "test",
+		WorkerVersionStamp: &commonpb.WorkerVersionStamp{
+			BuildId:       v1,
+			UseVersioning: true,
```
Comment (dnr) [inline tests/versioning_test.go:831]: weird to respond with UseVersioning if the poll didn't use versioning, but I guess it makes the test easier?

**Review by dnr — APPROVED** [review]

File: tests/versioning_test.go (hunk 2 above)
Comment (bergundy) [reply tests/versioning_test.go:831]: History doesn't care about version sets so I figured I could simplify the test code.

File: service/history/workflow_task_handler.go (hunk 1 above)
Comment (bergundy) [reply service/history/workflow_task_handler.go:397]: Sure, I can refactor

**Pull request merged**